File: birthday_problem/birthday_problem.js
// returns # of people until 2 people share birthday

function getRandomInt(min, max) {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min) + min) //The maximum is exclusive and the minimum is inclusive
}

function getRandomBirthday() {
  return getRandomInt(1, 366)
}

function simulate() {
  const birthdays = []

  for (let i = 1; i <= 365; i++) {
    const birthday = getRandomBirthday()
    if (birthdays.includes(birthday)) {
      return i
    }
    birthdays.push(birthday)
  }

  throw Error('Expected overlapping birthdays')
}

function main() {
  const simCount = 1_000_000
  const results = new Array(simCount)
    .fill(0)
    .map(simulate)
    .sort((a, b) => a - b)

  const mean = results.reduce((a, b) => a + b, 0) / results.length
  console.log(`mean: ${mean}, N=${simCount}`)
}
main()

File: birthday_problem/birthday_problem.c
#include <stdlib.h>
#include <time.h>
#include <stdio.h>
#include <math.h>

int random_range(int min, int max) {
  return ( rand() % ( max - min ) ) + min;
}

int get_random_birthday() {
  return random_range(1, 366);
}

int has_element(int a[], int sz, int x) {
  for (int i = 0; i < sz; i++) {
    if (a[i] == x) {
      return 1;
    }
  }
  return 0;
}

int simulate() {
  int birthdays[365] = {0};

  for (int i = 1; i <= 365; i++) {
    int birthday = get_random_birthday();
    if (has_element(birthdays, 365, birthday)) {
      return i;
    }
    birthdays[i-1] = birthday;
  }
  printf("Error: expected overlapping birthdays");
  return -1;
}

int main() {
  srand(time(NULL));
  
  int sim_count = 1000000;
  int results[sim_count];

  int sum = 0;
  float mean = 0.0;

  for (int i = 0; i < sim_count; i++) {
    results[i] = simulate();
    sum = sum + results[i];
  }


  mean = sum / sim_count;

  printf("mean: %f, N=%d\n", mean, sim_count);
  return 0;
}
